pandas/_libs/lib.pyx

```diff
@@ -2189,23 +2243,25 @@ def maybe_convert_numeric(
         int maybe_int
         Py_ssize_t i, n = values.size
         Seen seen = Seen(coerce_numeric)
-        ndarray[float64_t, ndim=1] floats = cnp.PyArray_EMPTY(
-            1, values.shape, cnp.NPY_FLOAT64, 0
-        )
+        # Okay to have this since it won't get allocated
+        # until a write happens
         ndarray[complex128_t, ndim=1] complexes = cnp.PyArray_EMPTY(
             1, values.shape, cnp.NPY_COMPLEX128, 0
         )
-        ndarray[int64_t, ndim=1] ints = cnp.PyArray_EMPTY(
-            1, values.shape, cnp.NPY_INT64, 0
-        )
-        ndarray[uint64_t, ndim=1] uints = cnp.PyArray_EMPTY(
-            1, values.shape, cnp.NPY_UINT64, 0
-        )
-        ndarray[uint8_t, ndim=1] bools = cnp.PyArray_EMPTY(
+        # We need to convert everything to complex before this
+        Py_ssize_t first_complex_idx = -1
+        # Keeps track of what element type is stored in each index
+        # Also tracks nulls for masked arrays
+        # (We'll fix mask if convert_to_masked_nullable=True)
+        # See convert_union_arr_to_ndarray
+        ndarray[uint8_t, ndim=1] mask = cnp.PyArray_EMPTY(
             1, values.shape, cnp.NPY_UINT8, 0
         )
-        ndarray[uint8_t, ndim=1] mask = np.zeros(n, dtype="u1")
+        NonComplexNumeric *arr = <NonComplexNumeric *>PyMem_Malloc(
+            n * sizeof(NonComplexNumeric)
```

Comment:
A union takes up the size of its largest element, so this doesn't really save anything over the previous approach

I'm also a bit wary of the PyMem_Malloc because it seems like there are quite a few code paths that would return with a memory leak

Reply:
From my understanding, the previous code allocated separate arrays for ints, floats, bools, etc, and did something like this:

See complex num: update complexes[i]
See float: update floats[i] and complexes[i]
See int: update floats[i], complexes[i]
See bool: update ints[i], floats[i], complexes[i]

So by using 1 array, we would save quite a bit for ints and bools in terms of peak memory usage, but not really that much for floats.

When trying to optimize this, I originally tried to use an object array to store elements and astype, but it turns out that that resulted in a regression with floats ( both memory-wise and speed-wise), which I think is caused by the fact that PyObjects are big in terms of byte sizes.

Reply:
Ah OK yea that is really inefficient. Nice find. And yea I wouldn't expect the PyObject case to necessarily be faster since you are going from an array of primitive types to an array of PyObject* that still point elsewhere in the heap for value retrieval

That said, I'm not sure about this resolution. I think you'd be better served refactoring this function and trying to dispatch to fused types when we know the required type

Reply:
> I think you'd be better served refactoring this function and trying to dispatch to fused types when we know the required type

IIUC the suggestion is a bit like infer_dtype or maybe_convert_objects where we iterate until we see a non-numeric entry and then dispatch based on that entry's type.  if we can use that pattern for the numeric types too that'd be great.

Reply:
If I'm not mistaken, ``maybe_convert_objects`` has the exact same pattern as ``maybe_convert_numeric``. 

The problem here is that unless we see a complex, we can't be sure of the return type until after 1 full pass, cause an element at the end could force an upcast.

The alternative to this method would be to iterate a second time through the array, and convert based on mask values set by the first pass, so we don't have to infer again but this may have a perf penalty (probably would result in better memory usage since we wouldn't have an intermediate array).

Reply:
Can you not just astype to a complex dtype when/if you come across it?